In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [17]:
data = pd.read_csv("../data/data.csv")
data = data.rename(columns={"E1": "E1 [GeV]", "E2": "E2 [GeV]", "px1 ": "px1 [GeV]", "px2": "px2 [GeV]", "py1": "py1 [GeV]", "py2": "py2 [GeV]", "pz1": "pz1 [GeV]", "pz2": "pz2 [GeV]", "phi1": "phi1 [rad]", "phi2": "phi2 [rad]", "pt1": "pt1 [GeV]", "pt2": "pt2 [GeV]", "M": "M [GeV]"})
data.head(10)

Run      Event  E1 [GeV]  px1 [GeV]  py1 [GeV]  pz1 [GeV]  pt1 [GeV]  \
0  147115  366639895  58.71410   -7.31132  10.531000  -57.29740   12.82020   
1  147115  366704169   6.61188   -4.15213  -0.579855   -5.11278    4.19242   
2  147115  367112316  25.54190  -11.48090   2.041680   22.72460   11.66100   
3  147115  366952149  65.39590    7.51214  11.887100   63.86620   14.06190   
4  147115  366523212  61.45040    2.95284 -14.622700  -59.61210   14.91790   
5  147115  366663412   6.39616   -5.45672  -2.090680   -2.60078    5.84352   
6  147115  366639101  84.50580    8.82436  10.578900   83.37530   13.77610   
7  147115  367133576  77.00570   10.00290   9.175450  -75.80000   13.57370   
8  147115  368639137   9.68787    1.11192   2.050640   -9.40284    2.33270   
9  147115  367825395  27.88120   11.93900 -18.346200   17.26960   21.88880   

       eta1  phi1 [rad]  Q1  E2 [GeV]  px2 [GeV]  py2 [GeV]  pz2 [GeV]  \
0 -2.202670    2.177660   1  11.28360  -1.032340   -1.88066  -11.07780   
1 -1.028420   -3.002840  -1  17.14920 -11.713500    5.04474   11.46470   
2  1.420480    2.965600   1  15.82030  -1.472800    2.25895  -15.58880   
3  2.218380    1.007210   1  25.12730   4.087860    2.59641   24.65630   
4 -2.093750   -1.371540  -1  13.88710  -0.277757   -2.42560  -13.67080   
5 -0.431551   -2.775710  -1  21.38650  15.169800   -8.87030  -12.18930   
6  2.500320    0.875576   1  12.67840  -1.134460   -3.20939  -12.21280   
7 -2.421030    0.742282   1   9.11623  -1.722950   -1.48674   -8.82761   
8 -2.102180    1.073940   1  63.45970  -1.858440   12.79460  -62.12870   
9  0.724032   -0.993887   1  12.92180  -5.026300   11.60260    2.66263   

   pt2 [GeV]      eta2  phi2 [rad]  Q2   M [GeV]  
0    2.14537 -2.344030   -2.072810  -1   8.94841  
1   12.75360  0.808077    2.734920   1  15.89300  
2    2.69667 -2.455080    2.148570   1  38.38770  
3    4.84272  2.330210    0.565865  -1   3.72862  
4    2.44145 -2.423700   -1.684810  -1   2.74718  
5   17.57280 -0.647450   -0.529120  -1  18.40230  
6    3.40400 -1.989560   -1.910570  -1  65.32390  
7    2.27574 -2.064940   -2.429650   1  11.29120  
8   12.92890 -2.273550    1.715040   1   3.58678  
9   12.64450  0.209050    1.979600  -1  34.26850

## Data Description
- Run: The run number of the event.
- Event: The event number.
- E1, E2: The total energy of the electron (GeV) for electrons 1 and 2.
- px1,py1,pz1,px2,py2,pz2: The components of the momentum of the electron 1 and 2 (GeV).
- pt1, pt2: The transverse momentum of the electron 1 and 2 (GeV).
- eta1, eta2: The pseudorapidity of the electron 1 and 2.
- phi1, phi2: The phi angle of the electron 1 and 2 (rad).
- Q1, Q2: The charge of the electron 1 and 2.
- M: The invariant mass of two electrons (GeV).


In [18]:
data.shape

(100000, 19)

## Cleaning the data

Check how many rows are unique based on the Event number.

In [19]:
data.value_counts(subset="Event") # There are 99976 unique rows

# If you want to see which rows are duplicates, run this code:
# duplicateCheck = data.duplicated(subset=["Event"], keep="first")
# print(data[duplicateCheck])

Event
440012942    3
439950319    3
519135297    2
542302783    2
542369397    2
            ..
264030150    1
264011215    1
264010599    1
264009979    1
264072987    1
Name: count, Length: 99976, dtype: int64

Deleting the duplicated rows from the original data.

In [20]:
data = data.drop_duplicates(subset=["Event"], keep="first")
data

Run      Event  E1 [GeV]  px1 [GeV]  py1 [GeV]  pz1 [GeV]  \
0      147115  366639895  58.71410  -7.311320  10.531000  -57.29740   
1      147115  366704169   6.61188  -4.152130  -0.579855   -5.11278   
2      147115  367112316  25.54190 -11.480900   2.041680   22.72460   
3      147115  366952149  65.39590   7.512140  11.887100   63.86620   
4      147115  366523212  61.45040   2.952840 -14.622700  -59.61210   
...       ...        ...       ...        ...        ...        ...   
99995  146511  522575834  12.31310 -10.658000   5.164440    3.36858   
99996  146511  522786431  18.46420   7.854990  15.133000   -7.08659   
99997  146511  522906124   4.18566  -3.273500  -0.308507   -2.59013   
99998  146511  523243830  54.46220  11.352600  11.880900   51.92400   
99999  146511  524172389   7.64000   0.886162   5.478900   -5.25033   

       pt1 [GeV]      eta1  phi1 [rad]  Q1  E2 [GeV]  px2 [GeV]  py2 [GeV]  \
0       12.82020 -2.202670    2.177660   1  11.28360  -1.032340   -1.88066   
1        4.19242 -1.028420   -3.002840  -1  17.14920 -11.713500    5.04474   
2       11.66100  1.420480    2.965600   1  15.82030  -1.472800    2.25895   
3       14.06190  2.218380    1.007210   1  25.12730   4.087860    2.59641   
4       14.91790 -2.093750   -1.371540  -1  13.88710  -0.277757   -2.42560   
...          ...       ...         ...  ..       ...        ...        ...   
99995   11.84330  0.280727    2.690370  -1   1.80181   0.668609   -1.58437   
99996   17.05020 -0.404510    1.092010   1  14.69110  -1.418020   -2.28117   
99997    3.28801 -0.723075   -3.047630   1  72.81740 -11.074900   -9.28179   
99998   16.43280  1.867800    0.808132  -1   8.58671   0.378009    3.07828   
99999    5.55010 -0.842662    1.410440   1  52.10880  16.807500   -4.60510   

       pz2 [GeV]  pt2 [GeV]      eta2  phi2 [rad]  Q2   M [GeV]  
0     -11.077800    2.14537 -2.344030   -2.072810  -1   8.94841  
1      11.464700   12.75360  0.808077    2.734920   1  15.89300  
2     -15.588800    2.69667 -2.455080    2.148570   1  38.38770  
3      24.656300    4.84272  2.330210    0.565865  -1   3.72862  
4     -13.670800    2.44145 -2.423700   -1.684810  -1   2.74718  
...          ...        ...       ...         ...  ..       ...  
99995   0.537805    1.71967  0.307851   -1.171470   1   8.44779  
99996 -14.443500    2.68598 -2.383880   -2.126960   1  20.71540  
99997 -71.369300   14.45010 -2.300410   -2.444050  -1  12.71350  
99998   8.007050    3.10141  1.677170    1.448610   1   4.69670  
99999  49.108400   17.42700  1.759250   -0.267427  -1  36.50430  

[99976 rows x 19 columns]

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99976 entries, 0 to 99999
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Run         99976 non-null  int64  
 1   Event       99976 non-null  int64  
 2   E1 [GeV]    99976 non-null  float64
 3   px1 [GeV]   99976 non-null  float64
 4   py1 [GeV]   99976 non-null  float64
 5   pz1 [GeV]   99976 non-null  float64
 6   pt1 [GeV]   99976 non-null  float64
 7   eta1        99976 non-null  float64
 8   phi1 [rad]  99976 non-null  float64
 9   Q1          99976 non-null  int64  
 10  E2 [GeV]    99976 non-null  float64
 11  px2 [GeV]   99976 non-null  float64
 12  py2 [GeV]   99976 non-null  float64
 13  pz2 [GeV]   99976 non-null  float64
 14  pt2 [GeV]   99976 non-null  float64
 15  eta2        99976 non-null  float64
 16  phi2 [rad]  99976 non-null  float64
 17  Q2          99976 non-null  int64  
 18  M [GeV]     99891 non-null  float64
dtypes: float64(15), int64(4)
memor

In [22]:
data = data.dropna(axis=0)
data

Run      Event  E1 [GeV]  px1 [GeV]  py1 [GeV]  pz1 [GeV]  \
0      147115  366639895  58.71410  -7.311320  10.531000  -57.29740   
1      147115  366704169   6.61188  -4.152130  -0.579855   -5.11278   
2      147115  367112316  25.54190 -11.480900   2.041680   22.72460   
3      147115  366952149  65.39590   7.512140  11.887100   63.86620   
4      147115  366523212  61.45040   2.952840 -14.622700  -59.61210   
...       ...        ...       ...        ...        ...        ...   
99995  146511  522575834  12.31310 -10.658000   5.164440    3.36858   
99996  146511  522786431  18.46420   7.854990  15.133000   -7.08659   
99997  146511  522906124   4.18566  -3.273500  -0.308507   -2.59013   
99998  146511  523243830  54.46220  11.352600  11.880900   51.92400   
99999  146511  524172389   7.64000   0.886162   5.478900   -5.25033   

       pt1 [GeV]      eta1  phi1 [rad]  Q1  E2 [GeV]  px2 [GeV]  py2 [GeV]  \
0       12.82020 -2.202670    2.177660   1  11.28360  -1.032340   -1.88066   
1        4.19242 -1.028420   -3.002840  -1  17.14920 -11.713500    5.04474   
2       11.66100  1.420480    2.965600   1  15.82030  -1.472800    2.25895   
3       14.06190  2.218380    1.007210   1  25.12730   4.087860    2.59641   
4       14.91790 -2.093750   -1.371540  -1  13.88710  -0.277757   -2.42560   
...          ...       ...         ...  ..       ...        ...        ...   
99995   11.84330  0.280727    2.690370  -1   1.80181   0.668609   -1.58437   
99996   17.05020 -0.404510    1.092010   1  14.69110  -1.418020   -2.28117   
99997    3.28801 -0.723075   -3.047630   1  72.81740 -11.074900   -9.28179   
99998   16.43280  1.867800    0.808132  -1   8.58671   0.378009    3.07828   
99999    5.55010 -0.842662    1.410440   1  52.10880  16.807500   -4.60510   

       pz2 [GeV]  pt2 [GeV]      eta2  phi2 [rad]  Q2   M [GeV]  
0     -11.077800    2.14537 -2.344030   -2.072810  -1   8.94841  
1      11.464700   12.75360  0.808077    2.734920   1  15.89300  
2     -15.588800    2.69667 -2.455080    2.148570   1  38.38770  
3      24.656300    4.84272  2.330210    0.565865  -1   3.72862  
4     -13.670800    2.44145 -2.423700   -1.684810  -1   2.74718  
...          ...        ...       ...         ...  ..       ...  
99995   0.537805    1.71967  0.307851   -1.171470   1   8.44779  
99996 -14.443500    2.68598 -2.383880   -2.126960   1  20.71540  
99997 -71.369300   14.45010 -2.300410   -2.444050  -1  12.71350  
99998   8.007050    3.10141  1.677170    1.448610   1   4.69670  
99999  49.108400   17.42700  1.759250   -0.267427  -1  36.50430  

[99891 rows x 19 columns]

In [23]:
data.shape

(99891, 19)

We are left with 99891 unique rows with no empty values.

In [25]:
data.to_csv("../data/electron.csv", index=False)